core/src/main/java/hudson/PluginWrapper.java

```diff
@@ -530,6 +564,95 @@ public void disable() throws IOException {
+                dependantsToCheck = Sets.difference(this.getDependants(), this.getOptionalDependants());
+        }
+
+        // Review all the dependants and add to the plugin result what happened with its dependants
+        for (String dependant : dependantsToCheck) {
+            PluginWrapper dependantPlugin = parent.getPlugin(dependant);
+
+            // The dependant plugin doesn't exist, add an error to the report
+            if (dependantPlugin == null) {
+                PluginDisableResult dependantStatus = new PluginDisableResult(dependant);
+                dependantStatus.setStatus(PluginDisableStatus.NO_SUCH_PLUGIN);
+                dependantStatus.setMessage(Messages.PluginWrapper_NoSuchPlugin(dependant));
+                result.addDependantDisableStatus(dependantStatus);
+
+            // If the strategy is none and there is some enabled dependant plugin, the plugin cannot be disabled. If
+            // this dependant plugin is not enabled, continue searching for one enabled.
+            } else if (strategy.equals(PluginDisableStrategy.NONE)) {
+                if (dependantPlugin.isEnabled()) {
+                    aDependantNotDisabled = dependant;
+                    break; // in this case, we don't need to continue with the rest of its dependants
+                }
+
+            // If the strategy is not none and this dependant plugin is not enabled, add it as already disabled
+            } else if (!dependantPlugin.isEnabled()) {
+                PluginDisableResult dependantStatus = new PluginDisableResult(dependant);
+                dependantStatus.setStatus(PluginDisableStatus.ALREADY_DISABLED);
+                dependantStatus.setMessage(Messages.PluginWrapper_Already_Disabled(dependant));
+                result.addDependantDisableStatus(dependantStatus);
+
+            // If the strategy is not none and this dependant plugin is enabled, disable it
+            } else {
+                // As there is no cycles in the plugin dependencies, the recursion shouldn't be infinite. The
+                // strategy used is the same for its dependants plugins
+                result.addDependantDisableStatus(dependantPlugin.disable(strategy));
+            }
+        }
+
+        // If there is no enabled-dependant plugin, disable this plugin and add it to the result
+        if (aDependantNotDisabled == null) {
+            this.disableWithoutCheck();
+            result.setMessage(Messages.PluginWrapper_Plugin_Disabled(shortName));
+            result.setStatus(PluginDisableStatus.DISABLED);
+
+        // if there is yet some not disabled dependant plugin (only possible with none strategy), this plugin cannot
+        // be disabled.
+        } else {
+            result.setMessage(Messages.PluginWrapper_Plugin_Has_Dependant(shortName, aDependantNotDisabled, strategy));
+            result.setStatus(PluginDisableStatus.NOT_DISABLED_DEPENDANTS);
+        }
+
+        return result;
+    }
```

Comment:
nit: This method is ~80 lines long, that seems a bit too much. Perhaps you can put some of the logic inside the different strategies (more OOP) to avoid mixing the behavior in this method.

Reply:
It's due to the comments, that I think are necessary. Anyway, I have encapsulated the dependantsToCheck in a method.